### PR TITLE
Multithreaded merge

### PR DESCRIPTION
**Pull request by michaelcfanning**

### CODE REVIEW

File: src/Test.UnitTests.Sarif.Multitool/MergeCommandTests.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ public class MergeCommandTestsFixture : DeletesOutputsDirectoryOnClassInitializa
         [Trait(TestTraits.Bug, "https://github.com/microsoft/sarif-sdk/issues/1592")]
         public void MergeCommand_WhenThereAreNoInputFiles_ProducesEmptyRunsArray()
         {
-            RunTest("NoInputFiles.sarif");
+//            RunTest("NoInputFiles.sarif");
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Multitool/MergeCommandTests.cs:37]: >NoInputFiles [](start = 23, length = 12)

Eddy, we need to uncomment this, probably a test failure to resolve here.

**Comment by eddynaka** [comment]
>             RunTest("NoInputFiles.sarif");

@michaelfanning , what would be the expected input for this?

---
Refers to: src/Test.UnitTests.Sarif.Multitool.Library/MergeCommandTests.cs:40 in c9f1ae6. [](commit_id = c9f1ae659f1d85bb99e7d421d2043f2706af5657, deletion_comment = False)

**Comment by eddynaka** [comment]
>             RunTest("NoInputFiles.sarif");

I think I already solved my question. No inputs at all. Thanks

---
In reply to: [716857042](https://github.com/microsoft/sarif-sdk/pull/2026#issuecomment-716857042) [](ancestors = 716857042)

---
Refers to: src/Test.UnitTests.Sarif.Multitool.Library/MergeCommandTests.cs:40 in c9f1ae6. [](commit_id = c9f1ae659f1d85bb99e7d421d2043f2706af5657, deletion_comment = False)

File: src/Sarif.Multitool.Library/MergeCommand.cs [hunk 2]
```diff
@@ -17,142 +21,251 @@ public class MergeCommand : CommandBase
...
-                if (mergeOptions.SplittingStrategy != SplittingStrategy.PerRule)
+                var workers = new Task<bool>[_options.Threads];
+                for (int i = 0; i < _options.Threads; i++)
                 {
-                    _fileSystem.DirectoryCreate(outputDirectory);
-
-                    // Write output to file.
-                    WriteSarifFile(_fileSystem, mergedLog, outputFilePath, mergeOptions.Formatting);
-                    return 0;
+                    workers[i] = Task.Run(LoadSarifLogs);
```
Comment (eddynaka) [inline src/Sarif.Multitool.Library/MergeCommand.cs:74]: I think I found the problem. Right now, we are creating multiple readers and I think it wasn't designed for that. Looking at the documentation, I saw the following example:
1 reader and 1 writer
1 reader and multiple writers

File: src/Test.UnitTests.Sarif.Multitool/MergeCommandTests.cs (hunk 1 above)
Comment (eddynaka) [reply src/Test.UnitTests.Sarif.Multitool/MergeCommandTests.cs]: Fixed!

File: src/Sarif.Multitool.Library/MergeCommand.cs (hunk 2 above)
Comment (eddynaka) [reply src/Sarif.Multitool.Library/MergeCommand.cs]: fixed. ignore my comment.

**Review by eddynaka — APPROVED** [review]

**Pull request merged**